The service-level fix looks right, and it matches what `assignProductVariantsToChannel` and `asset.service.ts` already do.

The two new e2e tests don't test what they claim to, though. By the time that `describe` block runs, an earlier test in the same file has already removed the shared `promotion` from `secondChannel`, so it only exists in the default channel. The restricted admin's session is on `secondChannel`, and both mutations look the promotion up via `findByIdsInChannel(ctx, Promotion, ids, ctx.channelId, {})`, so the promotion is invisible to it and the lookup comes back empty.

Revert the service change and run them — they do still fail, but like this:

```
AssertionError: promise resolved "{ assignPromotionsToChannel: [] }" instead of rejecting
```

The mutation never gets as far as the line that performs the cross-channel assign. All that's being exercised is the channel visibility filter.

To hit the actual path, the restricted admin needs a promotion it legitimately owns — created in or assigned to `secondChannel` — and then attempts to push that into `T_1`. That version genuinely succeeds before the fix and is rejected after it.

--- a/packages/core/e2e/promotion.e2e-spec.ts
+++ b/packages/core/e2e/promotion.e2e-spec.ts
@@ -401,6 +401,18 @@ describe('Promotion resolver', () => {
             beforeAll(async () => {
                 adminClient.setChannelToken(E2E_DEFAULT_CHANNEL_TOKEN);
                 await adminClient.asSuperAdmin();
+
+                // The restricted admin can only act on Promotions it can already see in its
+                // own active channel (secondChannel). Put the Promotion there first, as
+                // SuperAdmin, so the tests below exercise the real assign/remove logic
+                // instead of being stopped earlier by the channel-visibility filter.
+                await adminClient.query(assignPromotionsToChannelDocument, {
+                    input: {
+                        channelId: secondChannel.id,
+                        promotionIds: [promotion.id],
+                    },
+                });
+
                 const { createRole } = await adminClient.query(createRoleDocument, {
                     input: {
                         code: 'promotion-channel-permission-test-role',
@@ -425,6 +437,25 @@ describe('Promotion resolver', () => {
             afterAll(async () => {
                 await adminClient.asSuperAdmin();
                 adminClient.setChannelToken(E2E_DEFAULT_CHANNEL_TOKEN);
+                await adminClient.query(removePromotionsFromChannelDocument, {
+                    input: {
+                        channelId: secondChannel.id,
+                        promotionIds: [promotion.id],
+                    },
+                });
+            });
+
+            it('can assign a Promotion to a channel it has permission on', async () => {
+                // Positive control: proves the restricted admin's permissions genuinely work,
+                // so the rejections below can be trusted to be specifically about the target
+                // channel, not some other authorization gate.
+                const { assignPromotionsToChannel } = await adminClient.query(assignPromotionsToChannelDocument, {
+                    input: {
+                        channelId: secondChannel.id,
+                        promotionIds: [promotion.id],
+                    },
+                });
+                expect(assignPromotionsToChannel).toEqual([{ id: promotion.id, name: promotion.name }]);
             });
 
             it('cannot assign a Promotion to a channel it has no permission on', async () => {

--- a/packages/core/e2e/promotion.e2e-spec.ts
+++ b/packages/core/e2e/promotion.e2e-spec.ts
@@ -402,10 +402,8 @@ describe('Promotion resolver', () => {
                 adminClient.setChannelToken(E2E_DEFAULT_CHANNEL_TOKEN);
                 await adminClient.asSuperAdmin();
 
-                // The restricted admin can only act on Promotions it can already see in its
-                // own active channel (secondChannel). Put the Promotion there first, as
-                // SuperAdmin, so the tests below exercise the real assign/remove logic
-                // instead of being stopped earlier by the channel-visibility filter.
+                // Make the Promotion visible in secondChannel so the restricted admin can
+                // reach it below, rather than being stopped by the visibility filter.
                 await adminClient.query(assignPromotionsToChannelDocument, {
                     input: {
                         channelId: secondChannel.id,
@@ -446,9 +444,7 @@ describe('Promotion resolver', () => {
             });
 
             it('can assign a Promotion to a channel it has permission on', async () => {
-                // Positive control: proves the restricted admin's permissions genuinely work,
-                // so the rejections below can be trusted to be specifically about the target
-                // channel, not some other authorization gate.
+                // Positive control for the rejections below.
                 const { assignPromotionsToChannel } = await adminClient.query(assignPromotionsToChannelDocument, {
                     input: {
                         channelId: secondChannel.id,

--- a/packages/core/src/service/services/promotion.service.ts
+++ b/packages/core/src/service/services/promotion.service.ts
@@ -213,11 +213,9 @@ export class PromotionService {
         ctx: RequestContext,
         input: AssignPromotionsToChannelInput,
     ): Promise<Promotion[]> {
-        const hasPermission = await this.roleService.userHasPermissionOnChannel(
-            ctx,
-            input.channelId,
+        const hasPermission = await this.roleService.userHasAnyPermissionsOnChannel(ctx, input.channelId, [
             Permission.UpdatePromotion,
-        );
+        ]);
         if (!hasPermission) {
             throw new ForbiddenError();
         }
@@ -235,11 +233,9 @@ export class PromotionService {
     }
 
     async removePromotionsFromChannel(ctx: RequestContext, input: RemovePromotionsFromChannelInput) {
-        const hasPermission = await this.roleService.userHasPermissionOnChannel(
-            ctx,
-            input.channelId,
+        const hasPermission = await this.roleService.userHasAnyPermissionsOnChannel(ctx, input.channelId, [
             Permission.UpdatePromotion,
-        );
+        ]);
         if (!hasPermission) {
             throw new ForbiddenError();
         }

